plugin-server/src/main/ingestion-queues/analytics-events-ingestion-consumer.ts

```diff
@@ -46,3 +68,61 @@ export const startAnalyticsEventsIngestionConsumer = async ({
 
     return queue
 }
+
+export async function eachBatchIngestionWithOverflow(
+    payload: EachBatchPayload,
+    queue: IngestionConsumer
+): Promise<void> {
+    function groupIntoBatchesIngestion(kafkaMessages: KafkaMessage[], batchSize: number): KafkaMessage[][] {
+        // Once we see a distinct ID we've already seen break up the batch
+        const batches = []
+        const seenIds: Set<string> = new Set()
+        let currentBatch: KafkaMessage[] = []
+        for (const message of kafkaMessages) {
+            const pluginEvent = formPipelineEvent(message)
+            const seenKey = `${pluginEvent.team_id}:${pluginEvent.distinct_id}`
+            if (currentBatch.length === batchSize || seenIds.has(seenKey)) {
+                seenIds.clear()
+                batches.push(currentBatch)
+                currentBatch = []
+            }
+            seenIds.add(seenKey)
+            currentBatch.push(message)
+        }
+        if (currentBatch) {
+            batches.push(currentBatch)
+        }
+        return batches
+    }
+
+    await eachBatch(payload, queue, eachMessageIngestionWithOverflow, groupIntoBatchesIngestion, 'ingestion')
+}
+
+export async function eachMessageIngestionWithOverflow(message: KafkaMessage, queue: IngestionConsumer): Promise<void> {
+    const event = formPipelineEvent(message)
+
+    if (
+        // Events with a null key are produced to the the KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW topic, so we would
+        // never see them here, as this consumer's topic is set to KAFKA_EVENTS_PLUGIN_INGESTION.
+        // However, there could be some lingering events from before the new *_OVERFLOW topic was initialized.
+        // As we cannot accurately estimate capacity usage for those events, we will process them normally.
+        // Once lingering events are cleared up, this check is not be needed anymore except for type-checking.
+        message.key != null &&
+        ConfiguredLimiter.consume(message.key.toString(), 1) === false
+    ) {
```

Comment:
If the `key == null` it's fine for us to send it to overflow topic always, so we can simplify the code. 
```suggestion
    if (
        // Events with a null key are produced to the the KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW topic, so we would
        // never see them here, as this consumer's topic is set to KAFKA_EVENTS_PLUGIN_INGESTION. In any case any `null` key can be sent to overflow topic.
        message.key == null ||
        ConfiguredLimiter.consume(message.key.toString(), 1) === false
    ) {
```
if you choose to go with the above proposal, then here we could send events with `null` key to overflow topic & that would be fine for the historical events too.
```suggestion
    if (
        // Events are marked to have a null key during batch break-up if they should go to the overflow topic
        message.key == null
    ) {
```